com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs

```diff
@@ -345,34 +179,72 @@ void UpgradePrefabsAndPrefabInstances()
                     }
                 }
             }
+            
             m_CurrentSceneOrPrefab = string.Empty;
+        }
 
-            // In each scene, restore modifications in all prefab instances by copying data
-            // from the linked converted copy of the prefab instance.
-            
+        /// <summary>
+        /// Opens every scene in the project and upgrades:
+        /// 1. prefab instances using Conversion link,
+        /// 2. non-prefabs,
+        /// 3. cleans obsolete components,
+        /// 4. updates animation references.
+        /// </summary>
+        /// <param name="conversionLinksPerScene">A per scene (scene index) list of links between prefab instances
+        /// and their upgraded copies that contain the original prefab modifications.</param>
+        void UpgradeAllScenes(Dictionary<int, List<ConversionLink>> conversionLinksPerScene)
+        {
             for (var s = 0; s < m_SceneManager.SceneCount; ++s)
             {
+                // 0. Open scene
                 var scene = OpenScene(s);
                 var timelineManager = new TimelineManager(scene);
-                var conversionLinks = conversionLinksPerScene[m_SceneManager.GetScenePath(s)];
-                var allGameObjectsInScene = GetAllGameObjects();
+                var upgradables = GetUpgradeCandidates(scene.GetRootGameObjects());
```

Comment:
I think here is the answer to my other question.  The correct algorithm might have to be to do this whols thing in 2 passes: 
1. All the leaf objects that may be referenced (e.g. paths)
2. All the things that might refer to other upgraded things (e.g. vcams, freeLooks)

If in fact that is the solution, then we will lose the optimization because we'll have to double the number of scene opens

Reply:
I'll take a look